### PR TITLE
Rewrite Special:Browse and its factbox

### PR DESCRIPTION
**Pull request by alistair3149**
Follow up to #5786

This is required to standardize factbox across SMW.

- Separate data and HTML generation of Special:Browse factbox
- Switch Special:Browse factbox to Mustache
- Decouple and drop smw-table styles from factbox styles
- Clean up [Special:Browse HTML](https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/5788/files#diff-9663bdc7d036d8e19d4c8e035492e1097e701505a43123a09ca07e74553a9a3d)
- Clean up [factbox HTML](https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/5788/files#diff-877e3beef287fc455791d43e99ae9b85e74846697b8e5c9e265b2798baad7b89) classes and structure
- Simplify preloader behavior
- Clean up unused or duplicated factbox styles

https://github.com/user-attachments/assets/996e40b7-1a1c-402d-8e6a-fba9d111cd90

## Summary by CodeRabbit

- **New Features**
	- Introduced new styling for the browsing interface with a focus on a cleaner, more modern design.
	- Added new Mustache templates for rendering factboxes and properties, enhancing the visual structure of the output.

- **Bug Fixes**
	- Updated expected HTML class names in tests to reflect changes in the output structure, improving consistency.

- **Documentation**
	- Updated method documentation to reflect changes in return types and functionality for query form data retrieval.

- **Refactor**
	- Significant restructuring of CSS and JavaScript for the browsing module, enhancing maintainability and responsiveness.

- **Tests**
	- Updated test cases to align with new output structures and styling rules, ensuring accurate validation of functionality.

### CODE REVIEW

**Comment by alistair3149** [comment]
@coderabbitai review

File: templates/Factbox.mustache [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	</header>
+	{{/data-header}}
+	{{#array-sections}}{{>FactboxSection}}{{/array-sections}}
+	{{#data-pagination}}
+	<div class="smw-factbox-pagination">
+		{{{html-pagination-prev}}}
+		<div class="smw-factbox-pagination-status">{{{footer-content}}}</div>
+		{{{html-pagination-next}}}
+	</div>
+	{{/data-pagination}}
```
[the remark this answers is not in the document]
Comment (alistair3149) [reply templates/Factbox.mustache:15]: Will be followed up on another PR that converts pagination into actual buttons

[the remark this answers is not in the document]
Comment (alistair3149) [reply templates/Factbox.mustache:15]: No it is fine.

File: tests/phpunit/MediaWiki/Specials/SpecialBrowseTest.php [hunk 2]
```diff
@@ -85,17 +85,17 @@ public function queryParameterProvider() {
...
 			[
-				'data-subject="{&quot;dbkey&quot;:&quot;Main_Page&quot;,&quot;ns&quot;:0,&quot;iw&quot;:&quot;&quot;,&quot;subobject&quot;:&quot;_QUERY140d50d705e9566904fc4a877c755964&quot;}"',
-				'data-options="{&quot;dir&quot;:null,&quot;lang&quot;:&quot;en&quot;,&quot;group&quot;:null,&quot;printable&quot;:null,&quot;offset&quot;:null,&quot;including&quot;:null,&quot;showInverse&quot;:false,&quot;showAll&quot;:true,&quot;showGroup&quot;:false,&quot;showSort&quot;:false,&quot;api&quot;:true,&quot;valuelistlimit.out&quot;:&quot;30&quot;,&quot;valuelistlimit.in&quot;:&quot;20&quot;}"'
+				'data-mw-smw-browse-subject="{&quot;dbkey&quot;:&quot;Main_Page&quot;,&quot;ns&quot;:0,&quot;iw&quot;:&quot;&quot;,&quot;subobject&quot;:&quot;_QUERY140d50d705e9566904fc4a877c755964&quot;}"',
+				'data-mw-smw-browse-options="{&quot;dir&quot;:null,&quot;lang&quot;:&quot;en&quot;,&quot;group&quot;:null,&quot;printable&quot;:null,&quot;offset&quot;:null,&quot;including&quot;:null,&quot;showInverse&quot;:false,&quot;showAll&quot;:true,&quot;showGroup&quot;:false,&quot;showSort&quot;:false,&quot;api&quot;:true,&quot;valuelistlimit.out&quot;:&quot;30&quot;,&quot;valuelistlimit.in&quot;:&quot;20&quot;}"'
```
[the remark this answers is not in the document]
Comment (alistair3149) [reply tests/phpunit/MediaWiki/Specials/SpecialBrowseTest.php:98]: Those are unrelated to Special:Browse

File: src/MediaWiki/Specials/SpecialBrowse.php [hunk 3]
```diff
@@ -11,6 +11,7 @@
 use SMW\Message;
 use SMWInfolink as Infolink;
 use SpecialPage;
+use TemplateParser;
```
[the remark this answers is not in the document]
Comment (alistair3149) [reply src/MediaWiki/Specials/SpecialBrowse.php:14]: `TemplateParser` is the correct class.
It is moved to `MediaWiki\\Html\\TemplateParser` in later MW versions but `TemplateParser` is still working as an alias in MW 1.43.

File: src/MediaWiki/Specials/Browse/HtmlBuilder.php [hunk 4]
```diff
@@ -12,8 +12,8 @@
 use SMW\RequestOptions;
 use SMW\SemanticData;
 use SMW\Store;
-use SMW\Utils\HtmlDivTable;
 use SMWDataValue;
+use TemplateParser;
```
[the remark this answers is not in the document]
Comment (alistair3149) [reply src/MediaWiki/Specials/Browse/HtmlBuilder.php:16]: `TemplateParser` is the correct class.
It is moved to `MediaWiki\\Html\\TemplateParser` in later MW versions but `TemplateParser` is still working as an alias in MW 1.43.

File: templates/Factbox.mustache [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  Required data:
+  - is-loading: bool - Whether the factbox is loading
+  - data-header: object - Header information
+    - html-title: string - Title HTML
+    - html-actions: string - Actions HTML
+  - array-sections: object[] - Array of section data (rendered via FactboxSection partial)
+  - data-pagination: object - Pagination information
+    - html-pagination-prev: string - Previous page HTML
+    - html-pagination-next: string - Next page HTML
+    - footer-content: string - Pagination status HTML
```
Comment (Nikerabbit) [inline templates/Factbox.mustache:13]: Why this isn't prefixed with html-?

File: templates/SpecialBrowse.mustache [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	class="smw-browse"
+	{{#subject}}data-mw-smw-browse-subject="{{.}}"{{/subject}}
+	{{#options}}data-mw-smw-browse-options="{{.}}"{{/options}}
+>
+	{{#html-output}}{{{.}}}{{/html-output}}
+	{{#html-noscript}}<noscript>{{{.}}}</noscript>{{/html-noscript}}
+	<div class="smw-browse-content">
+		{{#data-factbox}}{{>Factbox}}{{/data-factbox}}
+		{{#data-form}}
+		<form class="smw-browse-search" name="smwbrowse" action="{{{form-action}}}" method="get" role="search">
```
Comment (Nikerabbit) [inline templates/SpecialBrowse.mustache:27]: Why are the attributes variables here and below not escaped?

File: templates/Factbox.mustache (hunk 5 above)
Comment (alistair3149) [reply templates/Factbox.mustache:13]: It is just a string and does not contain any HTML, but yes it should be escaped in Mustache

File: templates/SpecialBrowse.mustache (hunk 6 above)
Comment (alistair3149) [reply templates/SpecialBrowse.mustache]: Done

File: res/smw/factbox.less [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ *
+ * @since 5.0
+ *
+ * @file
+ * @ingroup SMW
+ *
+ * @licence GNU GPL v2+
+ * @author mwjames
```
Comment (JeroenDeDauw) [inline res/smw/factbox.less:26]: We finally found out who mwjames is!

**Comment by JeroenDeDauw** [comment]
I did a quick scan of this code and did not spot issues. Would be nice if someone else can verify the code is correc.

File: res/smw/factbox.less (hunk 7 above)
Comment (alistair3149) [reply res/smw/factbox.less:26]: I'm definitely not James haha, but I'm not sure what to do with the header so I just keep it there

Comment (JeroenDeDauw) [reply res/smw/factbox.less:26]: :detective: :thinking:

**Comment by alistair3149** [comment]
Converting to draft to see if I can add more tests

**Comment by JeroenDeDauw** [comment]
@SemanticMediaWiki/testers would be nice if someone could give this a spin

**Comment by alistair3149** [comment]
> @SemanticMediaWiki/testers would be nice if someone could give this a spin

Same! It is blocking the other UI-related patches and it would be awesome if we can move forward 🚀

**Comment by jaideraf** [comment]
Maybe a I can test it in the next week, but it would be nice if somebody else @SemanticMediaWiki/testers could do it.

**Comment by jaideraf** [comment]
Tested. All fine by me 👍🏼

**Pull request merged**